src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -22,40 +22,159 @@ import { first } from 'rxjs/operators';
 import {
   SavedObjectsClient,
   SavedObjectsSchema,
-  SavedObjectsRepository,
-  SavedObjectsSerializer,
   SavedObjectsClientProvider,
   ISavedObjectsClientProvider,
+  SavedObjectsClientProviderOptions,
 } from './';
-import { getRootPropertiesObjects } from './mappings';
 import { KibanaMigrator, IKibanaMigrator } from './migrations';
 import { CoreContext } from '../core_context';
-import { LegacyServiceSetup } from '../legacy/legacy_service';
-import { ElasticsearchServiceSetup } from '../elasticsearch';
+import { LegacyServiceDiscoverPlugins } from '../legacy/legacy_service';
+import { ElasticsearchServiceSetup, CallAPIOptions } from '../elasticsearch';
 import { KibanaConfigType } from '../kibana_config';
 import { retryCallCluster } from '../elasticsearch/retry_call_cluster';
 import { SavedObjectsConfigType } from './saved_objects_config';
 import { KibanaRequest } from '../http';
+import { SavedObjectsClientContract } from './types';
+import { ISavedObjectsRepository } from './service/lib/repository';
+import {
+  SavedObjectsClientFactory,
+  SavedObjectsClientWrapperFactory,
+} from './service/lib/scoped_client_provider';
+import { createRepository } from './service/lib/create_repository';
 import { Logger } from '..';
 
 /**
+ * Saved Objects is Kibana's data persisentence mechanism allowing plugins to
+ * use Elasticsearch for storing and querying state. The
+ * SavedObjectsServiceSetup API exposes methods for creating and registering
+ * Saved Object client wrappers.
+ *
+ * @remarks
+ * Note: The Saved Object setup API's should only be used for creating and
+ * registering client wrappers. Constructing a Saved Objects client or
+ * repository for use within your own plugin won't have any of the registered
+ * wrappers applied and is considered an anti-pattern. Use the Saved Objects
+ * client from the
+ * {@link SavedObjectsServiceStart | SavedObjectsServiceStart#scopedClient }
+ * method or the {@link RequestHandlerContext | route handler context} instead.
+ *
+ * When plugins access the Saved Objects client, a new client is created using
+ * the factory provided to `setClientFactory` and wrapped by all wrappers
+ * registered through `addClientWrapper`. To create a factory or wrapper,
+ * plugins will have to construct a Saved Objects client. First create a
+ * repository by calling `scopedRepository` or `internalRepository` and then
+ * use this repository as the argument to the {@link SavedObjectsClient}
+ * constructor.
+ *
+ * @example
+ * import {SavedObjectsClient, CoreSetup} from 'src/core/server';
+ *
+ * export class Plugin() {
+ *   setup: (core: CoreSetup) => {
+ *     core.savedObjects.setClientFactory(({request: KibanaRequest}) => {
+ *       return new SavedObjectsClient(core.savedObjects.scopedRepository(request));
+ *     })
+ *   }
+ * }
+ *
  * @public
  */
 export interface SavedObjectsServiceSetup {
-  clientProvider: ISavedObjectsClientProvider;
```

Comment:
Does this need to be removed from Core's exports?

Reply:
Although the type was called `SavedObjectsServiceSetup` it was only exposed internally so this belonged to what's now called `InternalSavedObjectsServiceSetup`, used only to expose it to legacy. I expose some of the clientProvider API from `SavedObjectsServiceSetup` through `setClientFactory` and `addClientWrapper`.

the `clientProvider` injected into legacy has now been moved to `InternalSavedObjectsServiceStart`. The reason I moved it into start was because legacy only uses this dependency in the start phase, but other than that this could be exposed in either since it's a legacy dependency that will fall away anyway.